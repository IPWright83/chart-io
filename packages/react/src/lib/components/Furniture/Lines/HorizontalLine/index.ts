import { withSVG } from "../../../../hoc";
import { HorizontalLine as HorizontalLineBase, IHorizontalLineProps } from "./HorizontalLine";

export const HorizontalLine = withSVG(HorizontalLineBase, "horizontal-line");

export type { IHorizontalLineProps };
