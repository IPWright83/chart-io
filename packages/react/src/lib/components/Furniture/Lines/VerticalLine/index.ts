import { withSVG } from "../../../../hoc";
import { VerticalLine as VerticalLineBase } from "./VerticalLine";

const VerticalLine = withSVG(VerticalLineBase, "vertical-line");

export { VerticalLine };
