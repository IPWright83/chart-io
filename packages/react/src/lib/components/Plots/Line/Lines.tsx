import { chartSelectors, IState } from "@chart-io/core";
import type { IPlotsProps } from "@chart-io/core";

import React from "react";
import { useSelector } from "react-redux";

import { Line } from "./Line";

export type ILinesProps = IPlotsProps;

/**
 * Represents a set of Line Plots
 * @param  {Object} props       The set of React properties
 * @return {ReactDOMComponent}  The Line plot component
 */
export function Lines({ ys, colors, ...props }: ILinesProps) {
    const theme = useSelector((s: IState) => chartSelectors.theme(s));
    const palette = colors || theme.series.colors;

    return (
        <React.Fragment>
            {ys.map((y, i) => (
                <Line {...props} key={y} y={y} color={palette[i]} />
            ))}
        </React.Fragment>
    );
}

Lines.requiresVirtualCanvas = false;
Lines.isPlot = true;
Lines.brush = {
    horizontal: true,
    vertical: false,
};
Lines.zoom = {
    horizontal: true,
    vertical: false,
};
