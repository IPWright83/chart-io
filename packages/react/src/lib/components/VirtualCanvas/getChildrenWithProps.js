import React from "react";

/**
 * Adds some additional props all the children
 * @param  {Array} options.children                     The children of the component
 * @param  {Function} options.renderVirtualCanvas       Function to call to update the virtual canvas
 * @return {Array}                                      The new children
 */
export const getChildrenWithProps = ({ children, renderVirtualCanvas }) => {
    if (children.length > 0) {
        return children.map((child) => {
            if (React.isValidElement(child)) {
                return React.cloneElement(child, { renderVirtualCanvas });
            }

            return child;
        });
    }

    if (React.isValidElement(children)) {
        return React.cloneElement(children, { renderVirtualCanvas });
    }

    return children;
};
