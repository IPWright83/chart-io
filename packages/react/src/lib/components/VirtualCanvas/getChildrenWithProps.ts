import type { IOnMouseOver, IOnMouseOut, IOnClick } from "@d3-chart/types";
import React from "react";

export type renderVirtualCanvasFunc = (
    update: d3.Transition<Element, unknown, any, unknown>,
    events: {
        onMouseOver: IOnMouseOver;
        onMouseOut: IOnMouseOut;
        onClick: IOnClick;
    }
) => void;

/**
 * Adds some additional props all the children
 * @param  children                    The children of the component
 * @param  renderVirtualCanvas         Function to call to update the virtual canvas
 * @return                             The new children
 */
export function getChildrenWithProps(children: any, renderVirtualCanvas: renderVirtualCanvasFunc): any {
    if (!children) {
        return children;
    }

    if (children.length > 0) {
        return children.map((child) => {
            if (React.isValidElement(child)) {
                // @ts-expect-error: We're dealing with really dynamic stuff here we can't really type against
                return React.cloneElement(child, { renderVirtualCanvas });
            }

            return child;
        });
    }

    if (React.isValidElement(children)) {
        // @ts-expect-error: We're dealing with really dynamic stuff here we can't really type against
        return React.cloneElement(children, { renderVirtualCanvas });
    }

    return children;
}
