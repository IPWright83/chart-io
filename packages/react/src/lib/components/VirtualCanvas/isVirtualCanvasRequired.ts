import type { ReactChildren } from "react";

/**
 * Determine if a Virtual Canvas is required or not
 * @param  children  The child or children of the virtual canvas
 * @return           True if a virtual canvas is required, otherwise false
 */
export const isVirtualCanvasRequired = (children: any): boolean => {
    if (!children) {
        return false;
    }

    // Array of children
    if (children.length) {
        return children.filter((c) => c?.type?.requiresVirtualCanvas).length > 0;
    }

    // Single child
    return children.requiresVirtualCanvas || children.type?.requiresVirtualCanvas;
};
