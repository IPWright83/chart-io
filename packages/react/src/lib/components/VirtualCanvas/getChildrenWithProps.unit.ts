import React from "react";

import { getChildrenWithProps } from "./getChildrenWithProps";

describe("VirtualCanvas getChildrenWithProps", () => {
    const renderVirtualCanvas = jest.fn();

    it("correctly augments props on a single child", () => {
        const mockElement = React.createElement("div");

        const child = getChildrenWithProps(mockElement, renderVirtualCanvas)[0];

        expect(child.type).toBe(mockElement.type);
        expect(child.props.renderVirtualCanvas).toBe(renderVirtualCanvas);
    });

    it("correctly augments a set of children", () => {
        const mockElement1 = React.createElement("div");
        const mockElement2 = React.createElement("p");

        const children = getChildrenWithProps([mockElement1, mockElement2], renderVirtualCanvas);

        expect(children.length).toBe(2);

        const child1 = children[0];
        expect(child1.type).toBe(mockElement1.type);
        expect(child1.props.renderVirtualCanvas).toBe(renderVirtualCanvas);

        const child2 = children[1];
        expect(child2.type).toBe(mockElement2.type);
        expect(child2.props.renderVirtualCanvas).toBe(renderVirtualCanvas);
    });

    it("correctly skips non React elements", () => {
        const notMockElement = {};

        const children = getChildrenWithProps([notMockElement], renderVirtualCanvas);

        expect(children.length).toBe(1);

        const child = children[0];
        expect(child).toBe(notMockElement);
        expect(child).toEqual({});
    });
});
