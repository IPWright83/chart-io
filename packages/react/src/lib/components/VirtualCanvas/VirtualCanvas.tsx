import { debounce } from "lodash";
import React, { useEffect, useCallback, useRef } from "react";
import { useStore, useSelector } from "react-redux";

import {
    addEventHandlers,
    clearVirtualCanvas,
    removeEventHandlers,
    renderVirtualCanvas,
    IColorToDataMap,
} from "../../hoc/canvas/virtual";
import { chartSelectors, IState } from "../../store";

import { getChildrenWithProps } from "./getChildrenWithProps";
import { isVirtualCanvasRequired } from "./isVirtualCanvasRequired";
import type { IRenderVirtualCanvasFunc } from "./getChildrenWithProps";

export const VIRTUAL_CANVAS_DEBOUNCE = 100;

export interface IVirtualCanvasProps {
    /**
     * The plots that are children of the virtual canvas
     */
    children: JSX.Element;
}

/**
 * The virtual canvas, draws elements to a non dom canvas and is used to
 * simulate mouse events on these elements
 * @param  {Object} props   The react props
 * @return {ReactElement}   A virtual canvas to add mouse events to canvas layers
 */
export function VirtualCanvas({ children }: IVirtualCanvasProps) {
    const width = useSelector((s: IState) => chartSelectors.dimensions.width(s));
    const height = useSelector((s: IState) => chartSelectors.dimensions.height(s));
    const store = useStore();

    // This is going to be used for the main color -> datum lookup.
    // We need to use a useRef so dependencies (wiring up events) don't re-occur forcing a re-render loop
    const colorToData = useRef<IColorToDataMap>({});
    const canvasRef = useRef<HTMLCanvasElement>();

    // We can't trigger a useEffect when the colorToData ref changes so
    // instead we use a callback function to ensure we always grab the
    // correct reference when trying to do our data lookups in the
    // addEventHandlers function
    const getColorMap = useCallback(() => colorToData.current, [colorToData.current]);

    // Adds event handlers to the canvas for triggering events
    useEffect(() => {
        const canvas = canvasRef.current;
        if (!canvas || !store) {
            return;
        }

        const { clickHandler, moveHandler } = addEventHandlers(canvas, getColorMap, store.dispatch);
        return () => {
            removeEventHandlers(canvas, clickHandler, moveHandler);
        };
    }, [canvasRef.current, store, getColorMap]);

    // This is a set of transient nodes, use as part of a debounced render function
    let nodes = [];

    // Render all the virtual nodes - this is debounced to ensure that we only trigger it once
    // after all of the child layers finished their render as we don't want layers to overwrite each other
    const renderAllVirtualNodes = debounce(async () => {
        const canvas = canvasRef.current;
        if (!canvas) {
            return;
        }

        // Clear and re-render the virtual canvas
        clearVirtualCanvas(canvas, width, height);
        colorToData.current = await renderVirtualCanvas(canvas, width, height, nodes);
        nodes = [];
    }, VIRTUAL_CANVAS_DEBOUNCE);

    // Whenever a child (canvas layer) renders it'll call this renderVirtual function
    // at the end of its render loop. We need to ensure that all nodes (virtual dom elements)
    // exist within the dataset and then render the virtual canvas
    const renderVirtual: IRenderVirtualCanvasFunc = (update, events) => {
        nodes = [
            ...nodes,
            {
                selection: update,
                events,
            },
        ];

        renderAllVirtualNodes();
    };

    // Many layers don't require the virtual canvas. If
    // they are all of these types then disable the canvas
    const includeVirtualCanvas = isVirtualCanvasRequired(children);
    if (includeVirtualCanvas === false) {
        return <React.Fragment>{children}</React.Fragment>;
    }

    // We need to extend the child components to provide the renderVirtual callback down
    // which will automatically be called by any Canvas based layers
    const childrenWithProps = getChildrenWithProps(children, renderVirtual);

    const style = {
        position: "absolute" as const,
        opacity: 0,
        zindex: 10000,
    };

    return (
        <React.Fragment>
            {childrenWithProps}
            <foreignObject className="fo" width={width} height={height}>
                <canvas className="virtual-canvas" width={width} height={height} ref={canvasRef} style={style}></canvas>
            </foreignObject>
        </React.Fragment>
    );
}
