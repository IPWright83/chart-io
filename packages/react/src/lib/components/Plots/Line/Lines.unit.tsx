import React from "react";
import { scaleLinear } from "@d3-chart/d3";
import { toMatchImageSnapshot } from "jest-image-snapshot";

import { VIRTUAL_CANVAS_DEBOUNCE, VirtualCanvas } from "../../VirtualCanvas";
import { Lines } from "./Lines";

expect.extend({ toMatchImageSnapshot });

import { getBuffer, renderChart, wait } from "../../../testUtils";

describe("Lines", () => {
    const data = [
        { y2: 5, y: 0, x: 0 },
        { y2: 10, y: 5, x: 1 },
        { y2: 0, y: 10, x: 2 },
    ];

    const scales = {
        y: scaleLinear().domain([0, 20]).range([100, 0]),
        y2: scaleLinear().domain([0, 20]).range([100, 0]),
        x: scaleLinear().domain([0, 5]).range([0, 100]),
    };

    describe("Multiple Series", () => {
        describe("using SVG", () => {
            it("should render correctly", async () => {
                const { asFragment } = await renderChart({
                    children: <Lines x="x" ys={["y", "y2"]} />,
                    data,
                    scales,
                });

                // Wait for the second render of the line, as
                // first render we put in a placeholder to animate
                await wait(10);

                expect(asFragment()).toMatchSnapshot();
            });
        });

        describe("using Canvas", () => {
            it("should render correctly", async () => {
                const { container } = await renderChart({
                    children: (
                        <VirtualCanvas>
                            <Lines x="x" ys={["y", "y2"]} useCanvas={true} />
                        </VirtualCanvas>
                    ),
                    data,
                    scales,
                });

                // Wait for the second render of the line, as
                // first render we put in a placeholder to animate
                await wait(10);

                await wait(VIRTUAL_CANVAS_DEBOUNCE * 2);

                const canvases = container.querySelectorAll(".canvas");
                expect(canvases.length).toBe(2);

                const canvasBuffer1 = getBuffer(canvases[0] as HTMLCanvasElement);
                expect(canvasBuffer1).toMatchImageSnapshot();

                const canvasBuffer2 = getBuffer(canvases[1] as HTMLCanvasElement);
                expect(canvasBuffer2).toMatchImageSnapshot();

                // const virtualCanvasBuffer = getBuffer(container.querySelector(".virtual-canvas"));
                // expect(virtualCanvasBuffer).toMatchImageSnapshot();
            });
        });
    });
});
