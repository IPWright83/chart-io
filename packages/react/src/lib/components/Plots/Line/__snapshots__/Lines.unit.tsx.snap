// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lines Multiple Series using SVG should render correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-plot line"
    >
      <path
        class="line"
        d="M0,100L20,75L40,50                                  "
        style="fill: none; pointer-events: none; stroke: rgb(153, 193, 220); stroke-width: 2px;"
      />
      <path
        class="line"
        d="M0,100L100,100"
        style="fill: none; pointer-events: none;"
      />
    </g>
    <g
      class="g-plot line"
    >
      <path
        class="line"
        d="M0,75L20,50L40,100                                  "
        style="fill: none; pointer-events: none; stroke: rgb(252, 153, 142); stroke-width: 2px;"
      />
      <path
        class="line"
        d="M0,100L100,100"
        style="fill: none; pointer-events: none;"
      />
    </g>
  </svg>
</DocumentFragment>
`;
